File: harness/ui/src/lib/docs.ts
import { Channel, invoke } from "@tauri-apps/api/core";
import { open } from "@tauri-apps/plugin-dialog";
import { createSignal } from "solid-js";
import { inTauri } from "./agent";
import { S, t, type Msg } from "./i18n";
import type {
  DocumentFormat,
  DocumentHit,
  DocumentView,
  IngestProgress,
  LibraryStats,
  OcrSetting,
} from "./protocol";

export type DocumentTaskKind = "sync" | "add" | "reprocess";
export type DocumentTaskState = "running" | "completed" | "failed";

export interface DocumentTaskFailure {
  path: string;
  error: string;
}

/** App-lifetime ingest state. Keeping this outside `DocsView` means switching screens does not lose a running task. */
export interface DocumentTask {
  id: string;
  scope: string;
  kind: DocumentTaskKind;
  state: DocumentTaskState;
  progress: IngestProgress;
  startedAt: number;
  finishedAt: number | null;
  stored: number;
  skipped: number;
  failures: DocumentTaskFailure[];
  warning: string | null;
  error: string | null;
  documents: DocumentView[] | null;
}

const [documentTasks, setDocumentTasks] = createSignal<Record<string, DocumentTask>>({});
const runningTasks = new Map<string, Promise<DocumentView[]>>();
let taskSequence = 0;

export { documentTasks };

function updateDocumentTask(
  scope: string,
  id: string,
  update: (task: DocumentTask) => DocumentTask,
) {
  setDocumentTasks((all) => {
    const current = all[scope];
    if (current === undefined || current.id !== id) return all;
    return { ...all, [scope]: update(current) };
  });
}

/** Run at most one document mutation per project while exposing its lifecycle to any mounted screen. */
export function runDocumentTask(
  scope: string,
  kind: DocumentTaskKind,
  initial: IngestProgress,
  execute: (onProgress: (progress: IngestProgress) => void) => Promise<DocumentView[]>,
): Promise<DocumentView[]> {
  const active = runningTasks.get(scope);
  if (active !== undefined) return active;

  const id = `${Date.now()}-${++taskSequence}`;
  const task: DocumentTask = {
    id,
    scope,
    kind,
    state: "running",
    progress: initial,
    startedAt: Date.now(),
    finishedAt: null,
    stored: 0,
    skipped: 0,
    failures: [],
    warning: null,
    error: null,
    documents: null,
  };
  setDocumentTasks((all) => ({ ...all, [scope]: task }));

  const note = (progress: IngestProgress) => {
    updateDocumentTask(scope, id, (current) => {
      const fatal = progress.stage === "failed" && progress.error !== null && progress.total === 0;
      const failure =
        progress.stage === "failed" && progress.error !== null && !fatal
          ? [{ path: progress.path, error: progress.error }]
          : [];
      return {
        ...current,
        progress,
        stored: current.stored + (progress.stage === "stored" ? 1 : 0),
        skipped: current.skipped + (progress.stage === "skipped" ? 1 : 0),
        failures: [...current.failures, ...failure],
        warning:
          progress.stage === "embedding" && progress.error !== null
            ? progress.error
            : current.warning,
        error: fatal ? progress.error : current.error,
      };
    });
  };

  const promise = execute(note)
    .then((documents) => {
      updateDocumentTask(scope, id, (current) => ({
        ...current,
        state: current.error === null ? "completed" : "failed",
        finishedAt: Date.now(),
        documents,
      }));
      return documents;
    })
    .catch((error: unknown) => {
      updateDocumentTask(scope, id, (current) => ({
        ...current,
        state: "failed",
        finishedAt: Date.now(),
        error: String(error),
      }));
      throw error;
    })
    .finally(() => {
      if (runningTasks.get(scope) === promise) runningTasks.delete(scope);
    });
  runningTasks.set(scope, promise);
  return promise;
}

export function dismissDocumentTask(scope: string) {
  setDocumentTasks((all) => {
    if (all[scope]?.state === "running") return all;
    const next = { ...all };
    delete next[scope];
    return next;
  });
}

export async function getOcrSetting(): Promise<OcrSetting> {
  if (!inTauri()) return { enabled: true, visionModel: null };
  try {
    return await invoke<OcrSetting>("ocr_setting");
  } catch (err) {
    console.error("failed to read OCR setting", err);
    return { enabled: true, visionModel: null };
  }
}

export function setOcrEnabled(enabled: boolean): Promise<OcrSetting> {
  return invoke<OcrSetting>("set_ocr_enabled", { enabled });
}

/** The document library commands. Screen-load calls swallow errors and return honest empty values; anything
 * behind a click throws, because silence after a click is indistinguishable from slowness. */

export async function listDocuments(): Promise<DocumentView[]> {
  if (!inTauri()) return [];
  try {
    return await invoke<DocumentView[]>("list_documents");
  } catch (err) {
    console.error("failed to list documents", err);
    return [];
  }
}

/** Library health; when unreachable it reports `semanticReady: false` with a reason, not stats that look empty. */
export async function libraryStats(): Promise<LibraryStats> {
  const unknown: LibraryStats = {
    documents: 0,
    chunks: 0,
    embeddedChunks: 0,
    embedder: null,
    semanticReady: false,
    reason: t(S.docs.error.stats),
    root: "",
    filesSeen: 0,
    filesSkipped: 0,
    unreadable: 0,
    excluded: 0,
    // `null`, not `0`: we do not know whether a scan ever ran, and `0` would render as 1/1/1970.
    scannedAt: null,
    scanning: null,
  };
  if (!inTauri()) return unknown;
  try {
    return await invoke<LibraryStats>("library_stats");
  } catch (err) {
    console.error("failed to read library stats", err);
    return unknown;
  }
}

/** Ingest documents, progress over a `Channel`; returns the library after ingest, and throws only if the whole batch fails. */
export function addDocuments(
  paths: string[],
  onProgress: (p: IngestProgress) => void,
): Promise<DocumentView[]> {
  const channel = new Channel<IngestProgress>();
  channel.onmessage = onProgress;
  return invoke<DocumentView[]>("add_documents", { paths, onProgress: channel });
}

/** Remove a document from the library, along with every chunk cut from it. */
/** Rescan the project directory on screen open, not on a button: the core skips unchanged files, so it is near free. */
export function syncLibrary(onProgress: (p: IngestProgress) => void): Promise<DocumentView[]> {
  if (!inTauri()) return Promise.resolve([]);
  const channel = new Channel<IngestProgress>();
  channel.onmessage = onProgress;
  return invoke<DocumentView[]>("sync_library", { onProgress: channel });
}

/** Reprocess the whole library after a click; throws rather than swallowing, unlike `syncLibrary`. */
export function reprocessLibrary(
  onProgress: (p: IngestProgress) => void,
): Promise<DocumentView[]> {
  const channel = new Channel<IngestProgress>();
  channel.onmessage = onProgress;
  return invoke<DocumentView[]>("reprocess_library", { onProgress: channel });
}

export function removeDocument(id: string): Promise<void> {
  return invoke("remove_document", { id });
}

/** Probe search: `limit` has a default because the probe answers "can the library find this", not "show me everything". */
export function searchDocuments(query: string, limit = 8): Promise<DocumentHit[]> {
  return invoke<DocumentHit[]>("search_documents", { query, limit });
}

/** OS file dialog; an empty array means the user cancelled. Unfiltered, since the readable-format list lives in Rust. */
export async function pickDocuments(): Promise<string[]> {
  if (!inTauri()) return [];
  const picked = await open({ directory: false, multiple: true });
  if (picked === null) return [];
  return Array.isArray(picked) ? picked : [picked];
}

/* --- Display helpers, so the document table does not invent its own naming --- */

export function formatLabel(format: DocumentFormat): string {
  return t(S.docs.format[format]);
}

/** Human file size, base 1024, one decimal from MB up, where the fraction is a real difference in load time. */
export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  const kb = bytes / 1024;
  if (kb < 1024) return `${Math.round(kb)} KB`;
  const mb = kb / 1024;
  if (mb < 1024) return `${mb.toFixed(1)} MB`;
  return `${(mb / 1024).toFixed(1)} GB`;
}

/** Three embedding states, kept out of JSX because both the table and the stats strip read them. */
export type EmbedState = "embedded" | "queued" | "failed";

/** `embedded === false` with `error === null` means queued, not failed; conflating the two makes users re-add good files. */
export function embedState(doc: DocumentView): EmbedState {
  if (doc.error !== null) return "failed";
  return doc.embedded ? "embedded" : "queued";
}

/** Translated ingest stage name; `stage` is a core key, and an unknown one is returned verbatim rather than blank. */
export function stageLabel(stage: string): string {
  const table: Record<string, Msg> = S.docs.stage;
  const msg = table[stage];
  return msg === undefined ? stage : t(msg);
}

File: harness/app/src/commands/docs.rs
//! The document library: ingest, list, delete, and test-search. No model tools appear here on purpose --
//! ingesting and deleting are user actions, and the model only gets `docs.search`, `docs.read` and
//! `docs.list`: it reads the library, it does not rearrange it.

use std::path::{Path, PathBuf};
use std::sync::Arc;

use futures::StreamExt;
use futures::stream::BoxStream;
use pai_rag::{DocLibrary, Docs, Document, IngestEvent};
use tauri::State;
use tauri::ipc::Channel;

use crate::AppState;
use crate::harness::Harness;
use crate::protocol::{DocumentHit, DocumentView, IngestProgress, LibraryStats, OcrSetting};

/// The open project's library; absence is valid, since code projects do not load `rag`, and the answer says
/// which project type this is rather than reporting a technical error.
fn library(harness: &Harness) -> Result<Arc<dyn DocLibrary>, String> {
    harness
        .ctx
        .get::<Docs>()
        .ok_or_else(|| "dự án đang mở không phải thư viện tài liệu".to_string())
}

fn view(doc: Document) -> DocumentView {
    DocumentView {
        id: doc.id,
        // The real path inside the project, not where the file came from: `origin` may have moved long ago.
        path: doc.path.display().to_string(),
        title: doc.title,
        format: doc.format.as_str().to_string(),
        bytes: doc.bytes,
        chunks: doc.chunks,
        pages: doc.pages,
        ocr_pages: doc.ocr_pages,
        embedded: doc.embedded,
        added_at: doc.added_at,
        error: doc.error,
    }
}

#[tauri::command]
pub async fn ocr_setting(state: State<'_, AppState>) -> Result<OcrSetting, String> {
    let harness = state.harness().await?;
    let (enabled, vision_model) = harness.rag_config.ocr_status();
    Ok(OcrSetting {
        enabled,
        vision_model,
    })
}

#[tauri::command]
pub async fn set_ocr_enabled(
    enabled: bool,
    state: State<'_, AppState>,
) -> Result<OcrSetting, String> {
    let harness = state.harness().await?;
    harness
        .rag_config
        .write_ocr_enabled(enabled)
        .map_err(|error| format!("không lưu được cấu hình OCR: {error}"))?;
    let (_, vision_model) = harness.rag_config.ocr_status();
    Ok(OcrSetting {
        enabled,
        vision_model,
    })
}

/// Drain an ingest stream, forwarding each milestone to the UI, then return the whole library; the three
/// commands below differ only in which stream they pass.
async fn drain(
    library: &Arc<dyn DocLibrary>,
    mut stream: BoxStream<'_, IngestEvent>,
    on_progress: Channel<IngestProgress>,
) -> Result<Vec<DocumentView>, String> {
    while let Some(event) = stream.next().await {
        let progress = IngestProgress {
            path: event.path,
            stage: event.stage.as_str().to_string(),
            done: event.done,
            total: event.total,
            finished: event.finished,
            error: event.error,
        };
        if let Err(err) = on_progress.send(progress) {
            // A closed window is no reason to abandon the ingest: files already copied into the store would be left half-done.
            tracing::debug!("could not send ingest progress: {err}");
        }
    }
    drop(stream);

    Ok(library
        .documents()
        .await
        .map_err(|err| err.to_string())?
        .into_iter()
        .map(view)
        .collect())
}

#[tauri::command]
pub async fn list_documents(state: State<'_, AppState>) -> Result<Vec<DocumentView>, String> {
    let harness = state.harness().await?;
    Ok(library(&harness)?
        .documents()
        .await
        .map_err(|err| err.to_string())?
        .into_iter()
        .map(view)
        .collect())
}

#[tauri::command]
pub async fn library_stats(state: State<'_, AppState>) -> Result<LibraryStats, String> {
    let harness = state.harness().await?;
    let stats = library(&harness)?
        .stats()
        .await
        .map_err(|err| err.to_string())?;
    Ok(LibraryStats {
        documents: stats.documents,
        chunks: stats.chunks,
        embedded_chunks: stats.embedded_chunks,
        embedder: stats.embedder,
        semantic_ready: stats.semantic_ready,
        reason: stats.reason,
        root: stats.root.display().to_string(),
        files_seen: stats.files_seen,
        files_skipped: stats.files_skipped,
        unreadable: stats.unreadable,
        excluded: stats.excluded,
        scanned_at: stats.scanned_at,
        scanning: stats.scanning.map(|item| crate::protocol::ScanProgress {
            done: item.done,
            total: item.total,
        }),
    })
}

/// Ingest a batch, emit progress, and return the whole library rather than the additions, since re-ingesting
/// updates existing rows. One bad file never fails the batch -- that is `pai-rag`'s contract, unchanged here.
#[tauri::command]
pub async fn add_documents(
    paths: Vec<String>,
    on_progress: Channel<IngestProgress>,
    state: State<'_, AppState>,
) -> Result<Vec<DocumentView>, String> {
    state.qdrant.ensure().await?;
    let harness = state.harness().await?;
    let library = library(&harness)?;
    let files: Vec<PathBuf> = paths.into_iter().map(PathBuf::from).collect();

    let stream = library.ingest(files);
    drain(&library, stream, on_progress).await
}

/// Rescan the project directory and sync the library with it, because the project directory is the library.
/// Not done inside `Library::open`, where a synchronous scan would freeze the window with nothing on screen.
#[tauri::command]
pub async fn sync_library(
    on_progress: Channel<IngestProgress>,
    state: State<'_, AppState>,
) -> Result<Vec<DocumentView>, String> {
    state.qdrant.ensure().await?;
    let harness = state.harness().await?;
    let library = library(&harness)?;

    let stream = library.sync();
    drain(&library, stream, on_progress).await
}

/// Reprocess the whole library on an explicit click. The automatic scan is incremental, so a file that failed
/// for a since-fixed reason is never revisited; this forgets every fingerprint, re-extracts, and embeds what
/// is missing. It is expensive, hence a button rather than automatic.
#[tauri::command]
pub async fn reprocess_library(
    on_progress: Channel<IngestProgress>,
    state: State<'_, AppState>,
) -> Result<Vec<DocumentView>, String> {
    state.qdrant.ensure().await?;
    let harness = state.harness().await?;
    let library = library(&harness)?;

    let stream = library.reprocess();
    drain(&library, stream, on_progress).await
}

/// Stop the current document ingest without rolling back files that were already committed.
#[tauri::command]
pub async fn stop_document_indexing(state: State<'_, AppState>) -> Result<bool, String> {
    let harness = state.harness().await?;
    Ok(library(&harness)?.cancel_ingest())
}

#[tauri::command]
pub async fn remove_document(id: String, state: State<'_, AppState>) -> Result<(), String> {
    let harness = state.harness().await?;
    library(&harness)?
        .remove(&id)
        .await
        .map_err(|err| err.to_string())
}

fn project_document_path(root: &Path, raw: &str) -> Result<PathBuf, String> {
    let root = root
        .canonicalize()
        .map_err(|err| format!("không mở được thư mục dự án: {err}"))?;
    let target = Path::new(raw)
        .canonicalize()
        .map_err(|err| format!("không mở được {raw}: {err}"))?;
    if target == root || !target.starts_with(&root) {
        return Err("tệp nằm ngoài thư mục dự án".to_string());
    }
    if !target.is_file() {
        return Err(format!("{} không phải là tệp", target.display()));
    }
    Ok(target)
}

/// Delete one file owned by the open document project and remove its derived chunks/vectors first. This is
/// deliberately separate from `remove_document`, whose Library-screen contract leaves the source file alone.
#[tauri::command]
pub async fn delete_project_document(
    path: String,
    state: State<'_, AppState>,
) -> Result<(), String> {
    let harness = state.harness().await?;
    let library = library(&harness)?;
    let root = harness
        .current_project()
        .map(|project| PathBuf::from(project.path))
        .ok_or_else(|| "chưa mở dự án".to_string())?;
    let target = project_document_path(&root, &path)?;

    let documents = library.documents().await.map_err(|err| err.to_string())?;
    if let Some(document) = documents.into_iter().find(|document| {
        document
            .path
            .canonicalize()
            .is_ok_and(|document_path| document_path == target)
    }) {
        library
            .remove(&document.id)
            .await
            .map_err(|err| err.to_string())?;
    }

    tokio::task::spawn_blocking(move || std::fs::remove_file(&target))
        .await
        .map_err(|err| format!("tác vụ xoá tệp bị dừng: {err}"))?
        .map_err(|err| format!("không xoá được {}: {err}", path))
}

/// A test search, so the user can verify the library before asking the assistant; matches come back verbatim
/// and the UI must render them as quotations, since this is third-party text.
#[tauri::command]
pub async fn search_documents(
    query: String,
    limit: Option<usize>,
    state: State<'_, AppState>,
) -> Result<Vec<DocumentHit>, String> {
    let harness = state.harness().await?;
    Ok(library(&harness)?
        .search(&query, limit.unwrap_or(8))
        .await
        .map_err(|err| err.to_string())?
        .into_iter()
        .map(|hit| DocumentHit {
            document_id: hit.document_id,
            title: hit.title,
            path: hit.path.display().to_string(),
            ordinal: hit.ordinal,
            text: hit.text,
            score: hit.score,
            matched_by: hit.matched_by.as_str().to_string(),
        })
        .collect())
}

#[cfg(test)]
mod tests {
    use super::project_document_path;

    #[test]
    fn project_document_must_be_a_file_inside_the_project() {
        let project = tempfile::tempdir().unwrap();
        let outside = tempfile::NamedTempFile::new().unwrap();
        let inside = project.path().join("notes.md");
        std::fs::write(&inside, "notes").unwrap();

        assert_eq!(
            project_document_path(project.path(), inside.to_str().unwrap()).unwrap(),
            inside.canonicalize().unwrap()
        );
        assert!(project_document_path(project.path(), outside.path().to_str().unwrap()).is_err());
        assert!(project_document_path(project.path(), project.path().to_str().unwrap()).is_err());
    }
}
